src/vulkan-renderer/application.cpp

```diff
@@ -473,10 +457,39 @@ Application::Application(int argc, char **argv) {
         enable_debug_marker_device_extension = false;
     }
 
-    m_device = std::make_unique<wrapper::Device>(*m_instance, m_surface->get(), enable_debug_marker_device_extension,
-                                                 use_distinct_data_transfer_queue, preferred_graphics_card);
+    const auto physical_devices = vk_tools::get_all_physical_devices(m_instance->instance());
+    if (preferred_graphics_card && *preferred_graphics_card >= physical_devices.size()) {
+        spdlog::critical("GPU index {} out of range!", *preferred_graphics_card);
+        throw std::runtime_error("Invalid GPU index");
+    }
+
+    const VkPhysicalDeviceFeatures required_features{
+        // Add required physical device features here
+    };
+
+    const VkPhysicalDeviceFeatures optional_features{
```

Comment:
Might be exposed at namespace level actually?

Reply:
I would just keep it there so the programmer can see which features are optional and which ones are required when creating the device. Putting it into some other place would make it a little harder to understand I suppose.